Update Link Findings to Services and Teams (#33468)
DataDog/documentation: content/en/security/code_security/software_composition_analysis/setup_static/_index.md

</div>

## Link findings to Datadog services and teams

Datadog associates code and library scan results with Datadog services and teams so that findings automatically route to the appropriate owners. This enables service-level visibility, ownership-based workflows, and faster remediation.

> To determine which service a vulnerability belongs to, Datadog evaluates several mapping mechanisms in the order listed below. **Each vulnerability is mapped with one method only:** if a mapping mechanism succeeds for a particular finding, Datadog does not attempt the remaining mechanisms for that finding.

Reviewer: ```suggestion
To determine the service a vulnerability belongs to, Datadog evaluates several mapping mechanisms in the order described in this section. 

**Each vulnerability is mapped with one method only:** if a mapping mechanism succeeds for a particular finding, Datadog does not attempt the remaining mechanisms for that finding.
```

<div class="alert alert-danger">

> Using service definitions that include code locations in the Software Catalog is the only way to explicitly control how static findings are mapped to services. The additional mechanisms described below—such as Error Tracking usage patterns and naming-based inference—are not user-configurable and depend on existing data from other Datadog products. Because of this, these mechanisms may not provide consistent mappings for organizations not using these products.

Reviewer: ```suggestion
Using service definitions that include code locations in the Software Catalog is the only way to explicitly control how static findings are mapped to services. The additional mechanisms described below, such as Error Tracking usage patterns and naming-based inference, are not user-configurable, and depend on existing data from other Datadog products. Consequently, these mechanisms might not provide consistent mappings for organizations not using these products.
```

</div>

### Mapping using the Software Catalog (recommended)

> Services in the Software Catalog can declare the parts of the codebase that belong to them using the `codeLocations` field. This field is available in the **Software Catalog [schema version `v3`][14]** and allows a service to specify:

Reviewer: ```suggestion
Services in the Software Catalog can declare the parts of the codebase that belong to them using the `codeLocations` field. This field is available in the Software Catalog [schema version `v3`][14] and allows a service to specify:
```

      paths:
        - path/to/service/code/**
    - repositoryURL: https://github.com/org/billing-service.git
      paths:
        - "**"
{{< /code-block >}}

The schema for this field is described in the [Software Catalog entity model][23].

Datadog will go through all your Software Catalog definitions and check whether your finding’s file path matches. For a finding to be mapped to a service through `codeLocations`, it must contain a file path. Some findings may not contain a file path; in those cases, Datadog cannot evaluate `codeLocations` for that finding, and this mechanism is skipped.

<div class="alert alert-danger">
Services defined with a **Software Catalog schema v2.x** do **not** support `codeLocations`. Existing definitions can be upgraded to the v3 schema in the [Software Catalog][24]. Once the migration is complete, changes may take up to 24 hours to apply to findings. If you are unable to upgrade to v3, Datadog will fall back to alternative linking techniques (described below). These rely on less precise heuristics, so accuracy may vary depending on the Code Security product and your use of other Datadog features.

> </div>

Reviewer: we can't use Markdown inside HTML div tags, only HTML.

#### Example (v3 schema)

{{< code-block lang="yaml" filename="entity.datadog.yaml" collapsible="true" >}}
apiVersion: v3
kind: service
metadata:
  name: billing-service
  owner: billing-team
datadog:
  codeLocations:
    - repositoryURL: https://github.com/org/myrepo.git
      paths:
        - path/to/service/code/**
    - repositoryURL: https://github.com/org/billing-service.git
      paths:
        - "**"
{{< /code-block >}}

#### SAST finding

> If a vulnerability appeared in `github.com/org/myrepo` at `/src/billing/models/payment.py`, then using the `codeLocations` for `billing-service` Datadog would add `billing-service` as an owning service. If your service defines an `owner` (see above), then Datadog will link that team to the finding too. In this case, the finding would be linked to the `billing-team`.

Reviewer: ```suggestion
If a vulnerability appeared in `github.com/org/myrepo` at `/src/billing/models/payment.py`, then using the `codeLocations` for `billing-service` makes Datadog add `billing-service` as an owning service. If your service defines an `owner` (see above), then Datadog links that team to the finding too. In this case, the finding would be linked to the `billing-team`.
```

Reviewer: @gorkavicente I'm not 100% sure I understand this one but I edited it for clarity. Can you confirm that it is accurate? thx

#### SCA finding

If a library was declared in `github.com/org/myrepo` at `/go.mod`, then Datadog would not match it to `billing-service`. 

If it was instead declared in `github.com/org/billing-service` at `/go.mod`, then Datadog would match it to `billing-service` due to the `“**”` catch-all glob. Consequently, Datadog would link the finding to the `billing-team`.

> **Datadog will attempt to map a single finding to as many services as possible. If no matches are found, Datadog will continue onto our next linking method.**

Reviewer: ```suggestion
Datadog attempts to map a single finding to as many services as possible. If no matches are found, Datadog continues onto our next linking method.
```

### When the Software Catalog cannot determine the service

> If the Software Catalog does not provide a match—either because the finding’s file path does not match any `codeLocations`, or because the service uses the v2.x schema—Datadog evaluates whether **Error Tracking** can identify the service associated with the code. Datadog uses only the last 30 days of Error Tracking data due to product [data-retention limits][25].

Reviewer: ```suggestion
If the Software Catalog does not provide a match, because the finding’s file path does not match any `codeLocations`, or because the service uses the v2.x schema, Datadog evaluates whether Error Tracking can identify the service associated with the code. Datadog uses only the last 30 days of Error Tracking data due to product [data-retention limits][25].
```

When Error Tracking processes stack traces, those traces often include **file paths**.
For example, if an error occurs in: `/foo/bar/baz.py`
Datadog inspects the directory: `/foo/bar`

> Datadog then checks whether the finding’s **file path** resides under that directory.

Reviewer: ```suggestion
Datadog then checks whether the finding’s file path resides under that directory.
```

**If the finding file is under the same directory:**

  - Datadog treats this as a strong indication that the vulnerability belongs to the same service
  - The finding inherits the **service** and **team** associated with that error in Error Tracking

**If this mapping succeeds, Datadog stops here.**

### Service inference from file paths or repository names

> When neither of the above strategies cannot determine the service, Datadog inspects naming patterns in the repository and file paths.

Reviewer: @gorkavicente just fixing the double negative of neither and cannot, but please check that this is still accurate

Datadog evaluates:

  - whether the **file path** contains identifiers matching a known service
  - whether the **repository name** corresponds to a service name

> When using the finding’s file path, Datadog performs a reverse search on each path segment until it finds a matching service or exhausts all options. For example, if a finding occurs in `github.com/org/checkout-service` at `/foo/bar/baz/main.go`, Datadog will take the last path segment, `main`, to see if any Software Catalog service uses that name. If a match exists, the finding is attributed to that service. If not, the process continues with `baz`, then `bar`, and so forth.

Reviewer: ```suggestion
When using the finding’s file path, Datadog performs a reverse search on each path segment until it finds a matching service or exhausts all options. For example, if a finding occurs in `github.com/org/checkout-service` at `/foo/bar/baz/main.go`, Datadog uses the last path segment, `main`, to determine if any Software Catalog service uses that name. If a match exists, the finding is attributed to that service. If not, the process continues with `baz`, then `bar`, and so on.
```

> If we have exhausted all options, Datadog will check whether the repository name, `checkout-service`, matches a Software Catalog service name. If no match was found, Datadog was unsuccessful at linking your finding using Software Catalog.

Reviewer: ```suggestion
When all options are exhausted, Datadog checks if the repository name, `checkout-service`, matches a Software Catalog service name. If no match is found, Datadog is unsuccessful at linking your finding using Software Catalog.
```

This mechanism ensures that findings still receive meaningful service attribution when no explicit metadata exists.

### Link findings to teams through Code Owners

> If Datadog was able to link your finding to a service using the above strategies, then the team that owns that service (if defined) would automatically be associated with that finding. Whether Datadog was able to successfully link a finding to a service (and a Datadog team) or not, Datadog will use the `CODEOWNERS` information from your finding’s repository to link Datadog and GitHub teams to your findings.

Reviewer: ```suggestion
If Datadog is able to link your finding to a service using the above strategies, then the team that owns that service (if defined) is automatically associated with that finding. Whether Datadog is able to successfully link a finding to a service (and a Datadog team) or not, Datadog uses the `CODEOWNERS` information from your finding’s repository to link Datadog and GitHub teams to your findings.
```
